### ISSUE TITLE
Incorrect offsets with add_prefix_space=True

### ISSUE THREAD

**Issue by david-waterworth** [huggingface/tokenizers#1300]
When I use `add_prefix_space=True` with added tokens, the char offset of the first token following an added token is off by one. i.e.

```
tokenizer = AutoTokenizer.from_pretrained("roberta-base", add_prefix_space=True)
tokenizer.add_tokens(["<device_name>", "</device_name>"], special_tokens=True)

encoding = tokenizer("<device_name>BMS-L17-1</device_name>")

for idx,token in enumerate(encoding.tokens()):
    print(token, encoding.token_to_chars(idx))
```

The second span should be `CharSpan(start=**13**, end=14)`

```
<s> None
<device_name> CharSpan(start=0, end=13)
ĠB CharSpan(start=14, end=14)
MS CharSpan(start=14, end=16)
- CharSpan(start=16, end=17)
L CharSpan(start=17, end=18)
17 CharSpan(start=18, end=20)
- CharSpan(start=20, end=21)
1 CharSpan(start=21, end=22)
</device_name> CharSpan(start=22, end=36)
</s> None
```

**Comment by Narsil** [comment]
This is unfortunate and resulting of old (but not really modifiable at this point) choices for this fast tokenizer.

As you see the token used is not `B` but `" B"`. Which means it has an extra space (in the token used).
This extra token shouldn't be in the offsets since it doesn't exist and is only created (correctly) by virtue of the tokenization scheme.

However the `start` gets removed 1 too much when trying to account for that.

I would love to be able to fix this, and PRs are welcome for it.
However at the current time, I don't really have the bandwidth to fix it.
I know fixing this is going to require touching core logic, that core logic is really hard to touch without destroying other corner cases. And unfortunately, this bug is relatively niche.

If anyone is interested in taking up on this. I'd be more than happy to provide guidance, but it's going to be tough work for sure.

**Comment by david-waterworth** [comment]
@Narsil Can I rely on the fact that the if the token starts with the added 'Ġ' character (and this character doesn't otherwise appear in the text), then the offset is always off by one? If this is the case, I can easily account for it in my code.

**Comment by arnaudstiegler** [comment]
Hi! Wanted to chime in, as I am facing a similar same issue with a tokenizer that I'm trying to fix.

Essentially, the tokenizer for the Donut model (RobertaTokenizer) is missing a token for "1" (that's just coming from the implementation apparently).
As a result:
```
from transformers import AutoTokenizer
​
tokenizer = AutoTokenizer.from_pretrained('nielsr/donut-base')
tokenizer.decode(tokenizer('A1B')['input_ids'])
# '<s> A<unk>B</s>'
```
As you can see, if there is a "1" directly preceded by a char, the token is not recognized and the fix would be to add "1" as a token, and that's where I hit the issue that is reported here.


```
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained('nielsr/donut-base')
tokenizer.add_tokens('1')
tokenizer.decode(tokenizer('A1B')['input_ids'])
# '<s> A1 B</s>'
```
Now the big issue is that I cannot safely remove those blanks because not every space between 1 and a char should be removed (imagine for instance an address).
This of course comes with the caveat that the issue is ultimately just an issue with the default tokenizer I'm using here

**Comment by ArthurZucker** [comment]
Hey! I'll probably look at this soon, as we fixed the `slow` version see #25224, were we have an extra space after special tokens. The issue is as expected, with `add_prefix` and splitting on the special tokens, a `prefix_space` is added before each word that comes up after a special/added token. 
@Narsil I'll investigate this 😉

**Comment by Narsil** [comment]
Making a general comment I always make about this kind of issue:

- Don't use `decode` if you can avoid it. Always use `offsets`. It's much more stable and avoids all the idiosyncracies that `decode` brings. (`decode` is really a best effort to represent what the input_ids are, but in general it's not a reversable operation).
